src/metaswitch/clearwater/queue_manager/queue_fsm.py

```diff
@@ -0,0 +1,138 @@
+                                                     self._set_timer_with_id],
+                           constants.LS_WAITING_ON_OTHER_NODE: [self._local_alarm.clear,
+                                                                self._set_timer_with_current_node_id],
+                           constants.LS_WAITING_ON_OTHER_NODE_ERROR: [self._local_alarm.critical,
+                                                                      self._set_timer_with_current_node_id]}
+
+        self._global_fsm = {constants.GS_NO_SYNC: self._global_alarm.clear,
+                            constants.GS_SYNC: self._global_alarm.minor,
+                            constants.GS_SYNC_ERROR: self._global_alarm.critical,
+                            constants.GS_NO_SYNC_ERROR: self._global_alarm.critical}
+
+    def quit(self):
+        if self._timer is not None:
+            self._timer.clear()
+        self._running = False
+
+    def is_running(self):
+        return self._running
+
+    def return_queue_config(self):
+        return self._queue_config
+
+    def fsm_update(self,
+                   queue_config):
+        """Main state machine function.
+
+        Arguments:
+            - queue_config_json: The current value in etcd
+
+        Returns:
+            - An updated queue_config_json
+              (which the caller can decide whether to write back to etcd)
+        """
+        self._queue_config = queue_config
+
+        # Parse the current queue configuration
+        self._queue_states = QueueStateInfo(self._id, self._queue_config)
+        _log.debug("Node at the front of the queue is %s" % self._queue_states.current_node_id())
+
+        # Firstly, check if we've entered the FSM because a timer has popped.
+        if self._timer is not None and self._timer.timer_popped:
+            _log.info("Timer has popped %s" % self._timer.timer_id)
+
+            # Does the timer relate to the node at the front of the queue?
+            if self._timer.timer_id == self._queue_states.current_node_id():
+                # Is the node at the front of the queue this node?
+                if self._queue_states.current_node_id() == self._id:
+                    self._local_alarm.critical()
+
+                mark_node_as_errored(self._queue_config, self._queue_states.current_node_id())
+                self._timer = None
+                return
+            else: #pragma: no cover
+                self._timer = None
+
+        # Now, check the local state and perform any appropriate actions
+        _log.debug("Local state is %s" % self._queue_states.local_state)
+        if self._last_local_state != self._queue_states.local_state:
+            for state_action in self._local_fsm[self._queue_states.local_state]:
+                self._last_local_state = self._queue_states.local_state
+                state_action()
+        
+        # Recalculate the global state, and raise the appropriate alarm given the
```

Comment:
Do we need to recalculate it? (What's changed _queue_config?)

If it needs recalculating, should QueueStateInfo expose a global_state() method that calculates it on the fly, or an explicit recalculate() method?